pkg/pillar/agentlog/agentlog_test.go

```diff
@@ -219,3 +234,897 @@ func TestPubsubLog(t *testing.T) {
+}
+
+func generateRandomAvgValue() float64 {
+	// Avg is a float of format 0.00 in range [0.00, 100.00]
+	seed := time.Now().UnixNano()
+	r := rand.New(rand.NewSource(seed))
+	return r.Float64() * 100
+}
+
+func matchPsiStats(line string) bool {
+	// The line should have the format "avg avg avg total avg avg avg total"
+	// where avg is a float of format 0.00 in range [0.00, 100.00] and total is an int from 0 to max uint64
+	re := regexp.MustCompile(`(\d{1,2}\.\d{2} ){3}\d{1,19} (\d{1,2}\.\d{2} ){3}\d{1,19}`)
+	return re.MatchString(line)
+}
+
+// Mutex for PSI stats producer - let's avoid running multiple producers at the same time
+var psiProducerMutex sync.Mutex
+
+func emulateMemoryPressureStats() (cancel context.CancelFunc, err error) {
+	// Take the mutex on the producer creation and release it when the producer is done
+	psiProducerMutex.Lock()
+	// Create a new file for memory pressure stats
+	fakePSIFileHandler, err := os.CreateTemp("", "memory-pressure")
+	if err != nil {
+		return nil, err
+	}
+
+	fakePSIFileName := fakePSIFileHandler.Name()
+	originalPressureMemoryFile := agentlog.PressureMemoryFile
+	agentlog.PressureMemoryFile = fakePSIFileName
+
+	// Start a ticker to write a new line to the file every 0.5 seconds
+	ticker := time.NewTicker(500 * time.Millisecond)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	go func() {
+		defer func() {
+			ticker.Stop()
+			agentlog.PsiMutex.Lock()
+			fakePSIFileHandler.Close()
+			os.Remove(fakePSIFileName)
+			agentlog.PressureMemoryFile = originalPressureMemoryFile
+			agentlog.PsiMutex.Unlock()
+			// We destroy this producer, so release the mutex
+			psiProducerMutex.Unlock()
+		}()
+		PsiStats := agentlog.PressureStallInfo{
+			SomeAvg10:  0.00,
+			SomeAvg60:  0.00,
+			SomeAvg300: 0.00,
+			SomeTotal:  0,
+			FullAvg10:  0.00,
+			FullAvg60:  0.00,
+			FullAvg300: 0.00,
+			FullTotal:  0,
+		}
+		for {
+			select {
+			case <-ticker.C:
+				agentlog.PsiMutex.Lock()
+				PsiStats.SomeAvg10 = generateRandomAvgValue()
+				PsiStats.SomeAvg60 = generateRandomAvgValue()
+				PsiStats.SomeAvg300 = generateRandomAvgValue()
+				PsiStats.SomeTotal = increaseTotalRandomly(PsiStats.SomeTotal)
+				PsiStats.FullAvg10 = generateRandomAvgValue()
+				PsiStats.FullAvg60 = generateRandomAvgValue()
+				PsiStats.FullAvg300 = generateRandomAvgValue()
+				PsiStats.FullTotal = increaseTotalRandomly(PsiStats.FullTotal)
+				// Rewrite the content of the file
+				content := fmt.Sprintf(
+					"some avg10=%.02f avg60=%.02f avg300=%.02f total=%d\nfull avg10=%.02f avg60=%.02f avg300=%.02f total=%d\n",
+					PsiStats.SomeAvg10, PsiStats.SomeAvg60, PsiStats.SomeAvg300, PsiStats.SomeTotal,
+					PsiStats.FullAvg10, PsiStats.FullAvg60, PsiStats.FullAvg300, PsiStats.FullTotal)
+				if err := os.WriteFile(fakePSIFileName, []byte(content), 0644); err != nil {
+					panic(err)
+				}
+				agentlog.PsiMutex.Unlock()
+			case <-ctx.Done():
+				return
```

Comment:
```suggestion
				break
```
and then you don't need `defer func()`

Reply:
You mean like this?
```go
			case <-ctx.Done():
				ticker.Stop()
				agentlog.PsiMutex.Lock()
				fakePSIFileHandler.Close()
				os.Remove(fakePSIFileName)
				agentlog.PressureMemoryFile = originalPressureMemoryFile
				agentlog.PsiMutex.Unlock()
				// We destroy this producer, so release the mutex
				psiProducerMutex.Unlock()
				break
```
But with return it will work as well?

Reply:
I tried and now my tests fail with double unlock

Reply:
ah, you're right, because `break` does not exit from the for-loop but from the switch

Reply:
yep, with `return` it works as expected...